protocol/x/vault/keeper/withdraw.go

```diff
@@ -0,0 +1,117 @@
+package keeper
+
+import (
+	"math/big"
+
+	errorsmod "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/dydxprotocol/v4-chain/protocol/lib"
+	"github.com/dydxprotocol/v4-chain/protocol/lib/vault"
+	"github.com/dydxprotocol/v4-chain/protocol/x/vault/types"
+)
+
+// GetVaultWithdrawalSlippagePpm returns the slippage (in ppm) that should be incurred
+// on withdrawing `withdrawalPortionPpm` of a vault's ownership.
+// For example, if `withdrawalPortionPpm = 100_000` and `200_000` is returned,
+// it means that withdrawing 10% has a 20% slippage for the specified vault.
+//
+// Slippage is calculated as `min(simple_slippage, estimated_slippage)` where:
+// - simple_slippage = leverage * initial_margin
+// - estimated_slippage = spread * (1 + average_skew) * leverage
+//   - average_skew = integral / (posterior_leverage - leverage)
+//   - integral = skew_antiderivative(skew_factor, posterior_leverage) -
+//     skew_antiderivative(skew_factor, leverage)
+//   - posterior_leverage = leverage / (1 - withdrawal_portion)
+//     = leverage / (1 - shares_to_withdraw / total_shares)
+//     = leverage * total_shares / (total_shares - shares_to_withdraw)
+func (k Keeper) GetVaultWithdrawalSlippagePpm(
+	ctx sdk.Context,
+	vaultId types.VaultId,
+	sharesToWithdraw *big.Int,
+) (*big.Int, error) {
```

Comment:
non-blocking, I wonder if it would be better to have this be a state-less function instead that takes in quoting params + leverage + total shares + shares to withdraw, rather than fetching from state.
Benefits:
- tests run a lot faster
- [vague] can run multiple times w/o fetching state multiple times, perhaps if we change input params
Cons:
- more indirection?
- more code?

Reply:
good thought. will keep this in mind when finishing up withdrawal and see if changing this to stateless makes sense